### DISCUSSION TITLE
Looking for performance improvements on job scheduling problem

### DISCUSSION THREAD

**Discussion by bradleybauer** [google/or-tools#3172] [CP-SAT questions]
I am using ortools to solve a job scheduling problem where the goal is to minimize the completion time. Is there anything I can do to improve the performance of the solver? Would it be worth it to build ortools from the master branch?

Here is a [link](https://github.com/bradleybauer/or) to the code. `main.py` has the solver definition. `VarManager` class helps create variables (all IntVars and BoolVars are created there). `Problem` contains static methods to provide a few sample problems.

Any help is greatly appreciated.

**Problem Description**
- No jobs are explicitly given. Instead a map of work types to work amounts is given.
- The amount of work for a work type is integer and is referred to as a number of runs.
- A job is an integer number of runs of some type scheduled to be processed serially.
- A job is made up of runs of only one work type.
- A job can contain a minimum of one run.
- A job can be said to have the same type as the work it is made from.
- Jobs are completed by scheduling them to machines to be processed.
- Jobs cannot be interrupted.
- There are two different parallel machines, machine 1 and machine 2.
- Machine 1 and machine 2 work in parallel.
- Machine `m` is said to have `maxNumSlotsPerJob[m]` different slots where each slot can processes one job at a time.
- Jobs on different slots are processed in parallel.
- Job types can be processed on either machine 1 or machine 2 but not both.
- Each job can occupy at most `maxNumSlotsPerJob` number of slots.

- There are dependencies between work of different types.
- Dependencies are not circular.
- If work type `k2` depends on work type `k1`, then a number of runs of type `k2` depend on some number of runs of type `k1`.
- The dependency relation has the form `NumRequired(k1,k2,num_k2_runs) = ceil(num_k2_runs * NumPerRun(k1,k2) * (1-machineBonusFloat))` where `machineBonusFloat` is in the interval `[0,.5]`.

    Note that `NumRequired` is not linear since it contains a `ceil`.
    This causes a few hiccups when calculating dependency constraints.
    Basically, the number of `k1` required by `k2`, on a particular batch,
    depends not only on the number of runs of `k2` but also on how the runs are allocated to the slots.
- The exact amount of work scheduled for types that are dependencies of other types is not specified.

    For example if work of type `k2` depend on work of type `k1` and `N` runs of type `k2` are specified
    in the problem input but runs of type `k1` are not specified then the amount of `k1` needed will
    be determined by `NumRequired(k1,k2,N)` (if all `k2` is on one slot). But if an amount `X` of `k1`
    is specified in the problem input then the total amount scheduled should be `X + NumRequired(k1,k2,N)`
    (if `k2` is the only type that depends on `k1`). In other words, the amount of work needed for dependency
    constraints is separate from the amount of work specified in the problem input.
- Machine 1 jobs can depend only on machine 1 jobs but machine 2 jobs can depend on machine 1 jobs or machine 2 jobs.

- Jobs scheduled to machine `m` are scheduled in batches.
- A batch is processed on either machine 1 or on machine 2.
- All jobs in a batch have the same starting time.
- The processing time of a batch is the maximum processing time among all jobs in the batch.
- When a job on a certain batch finishes, the slot that the job was using remains empty until the next batch where the slot is possibly used for some other job.
- When a machine finishes processing a batch it begins processing the next batch immediately.
- Scheduling is done in batches for two reasons:
    1. Schedules with smaller number of distinct starting times are easier to manage
       by the person using the schedule.
    2. I think the number of variables and dependency constraints becomes unmanageable if the
       number of distinct job starting times is allowed to get large.

- An inventory can be provided.

    An inventory is a map from work types to work amounts.
    This allows the user to supply intermediates so that dependency constraints can be relaxed and
    certain jobs can be skipped. For example, if the inventory contains `k1 -> 100` and `NumRequired(k1,k2,N)`
    demands that 101 runs of `k1` be processed, then after considering the inventory the total amount
    of `k1` needed by `k2` would be 1. Note that inventory is only used to satisfy dependency constraints.
    If `X` amount `k1` is specified in the problem input then atleast `X` runs of `k1` must be in the final schedule.

- The amount of time a job takes to complete has the form `JobTime(jobType, numRuns) = numRuns * TimePerRun(jobType)` where `TimePerRun(jobType)` is an integer.

**Comment by lperron** [comment]
the most expensive constraints are the modulo/division/product.

Can you expand on the equations that uses them. For instance, at some point, you compute the ceil of a division.

**Reply by bradleybauer** [reply]
The dependency calculations use `ceil(floor(r/max(1,s))c)*(s-(r%s)) + ceil((floor(r/max(1,s)) + 1)c)*(r%s)` for integers `r,s` and rational `c`. Unfortunately I do not see a way to simplify that even if I write `floor(r/s)` instead of `floor(r/max(1,s))`.

But, I was able to remove a few constraints by rewriting `ceil(x/y)` as `floor((x + y - 1) / y)`.
